test/jdk/java/lang/Thread/virtual/MonitorEnterWaitOOME.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test id=monitorenter
+ * @bug 8345266
+ * @summary Test OOM while trying to unmount vthread on monitorenter
+ * @requires vm.continuations & vm.gc.G1 & vm.opt.DisableExplicitGC != "true"
+ * @library /test/lib
+ * @run main/othervm -XX:+UseG1GC -Xmx48M MonitorEnterWaitOOME false
+ */
+
+/*
+ * @test id=timedwait
+ * @summary Test OOM while trying to unmount vthread on Object.wait
+ * @requires vm.continuations & vm.gc.G1 & vm.opt.DisableExplicitGC != "true"
+ * @library /test/lib
+ * @run main/othervm -XX:+UseG1GC -Xmx48M MonitorEnterWaitOOME true 5
+ */
+
+/*
+ * @test id=untimedwait
+ * @summary Test OOM while trying to unmount vthread on Object.wait
+ * @requires vm.continuations & vm.gc.G1 & vm.opt.DisableExplicitGC != "true"
+ * @library /test/lib
+ * @run main/othervm -XX:+UseG1GC -Xmx48M MonitorEnterWaitOOME true 0
+ */
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class MonitorEnterWaitOOME {
+    static volatile Object data;
+    static Thread.State dummyState = Thread.State.RUNNABLE; // load java.lang.Thread$State
+
+    public static void main(String[] args) throws Throwable {
+        final boolean testWait = args.length >= 1 ? Boolean.parseBoolean(args[0]) : false;
+        final long timeout = testWait && args.length == 2 ? Long.parseLong(args[1]) : 0L;
+
```

Comment:
At some point we might have to add VThreadRunner.ensureParallelism(2) to main to ensure there are at least two carriers. Just thinking of JTREG_TEST_THREAD_FACTORY=Virtual on one core systems, a configuration that seems to come sometimes with ATR.

Reply:
Added.